File: app/src/main/java/com/cyrillrx/uremote/explorer/Explorer.java
package com.cyrillrx.uremote.explorer;

import com.cyrillrx.android.utils.FileUtils;
import com.cyrillrx.uremote.request.protobuf.RemoteCommand;

import java.io.File;

/**
 * @author Cyril Leroux
 *         Created 12/05/2017.
 */
public abstract class Explorer {

    public static final String PREVIOUS_DIRECTORY_PATH = "..";
    public static final String KEY_DIRECTORY_CONTENT = "DIRECTORY_CONTENT";
    public static final String DEFAULT_DIRECTORY_PATH = "default_path";

    protected String root;
    protected String path;

    protected RemoteCommand.FileInfo currentFileInfo;

    protected Callback callback;

    public Explorer(Callback callback) {
        this.callback = callback;
        root = DEFAULT_DIRECTORY_PATH;
    }

    public boolean setRoot(String root) {

        if (root == null) {
            this.root = DEFAULT_DIRECTORY_PATH;
            return false;
        }

        this.root = root;
        navigateTo(root);
        return true;
    }

    public void reload() {
        if (path != null) {
            navigateTo(path);
        }
    }

    /**
     * Lists the content of the passed directory.<br />
     * Updates the view once the data have been received.
     *
     * @param dirPath The path of the directory to display.
     */
    public void navigateTo(String dirPath) { path = dirPath; }

    /**
     * Navigates up if possible.<br />
     * This method is supposed to be called from the parent Activity (most likely through the ActionBar).<br />
     * Updates the view once the data have been received from the server.
     *
     * @return True if we can navigate up from the current directory. False otherwise.
     */
    public boolean navigateUp() {

        if (canNavigateUp()) {
            doNavigateUp();
            return true;
        }
        return false;
    }

    /**
     * Call by the activity that holds the fragment if the back button is override.
     * If the function returns true, the back button is override to go up.
     * Else, it behaves normally.
     *
     * @return True if we can navigate up from the current directory. False otherwise.
     */
    public boolean canNavigateUp() {

        if (currentFileInfo == null) { return false; }

        final String absolutePath = currentFileInfo.getAbsoluteFilePath();

        return absolutePath != null &&
                !absolutePath.equals(root) &&
                absolutePath.contains(File.separator);

    }

    /**
     * Calls navigateTo on the parent directory.
     */
    protected void doNavigateUp() {
        final String parentPath = FileUtils.truncatePath(currentFileInfo.getAbsoluteFilePath());
        navigateTo(parentPath);
    }

    public interface Callback {
        void onDirectoryLoaded(RemoteCommand.FileInfo dirContent);
    }
}

File: app/src/main/java/com/cyrillrx/uremote/explorer/Explorer.kt
package com.cyrillrx.uremote.explorer

import com.cyrillrx.android.utils.FileUtils
import com.cyrillrx.uremote.request.protobuf.RemoteCommand

import java.io.File

/**
 * @author Cyril Leroux
 *         Created 12/05/2017.
 */
abstract class Explorer(protected var callback: Explorer.Callback) {

    protected var root: String
    protected var path: String

    protected var currentFileInfo: RemoteCommand.FileInfo? = null

    init {
        root = DEFAULT_DIRECTORY_PATH
        path = root
    }


    fun changeRoot(root: String): Boolean {

        this.root = root
        navigateTo(root)
        return true
    }

    fun reload() {
        navigateTo(path)
    }

    /**
     * Lists the content of the passed directory.<br></br>
     * Updates the view once the data have been received.
     *
     * @param dirPath The path of the directory to display.
     */
    open fun navigateTo(dirPath: String) {
        path = dirPath
    }

    /**
     * Navigates up if possible.<br></br>
     * This method is supposed to be called from the parent Activity (most likely through the ActionBar).<br></br>
     * Updates the view once the data have been received from the server.
     *
     * @return True if we can navigate up from the current directory. False otherwise.
     */
    fun navigateUp(): Boolean {

        if (canNavigateUp()) {
            doNavigateUp()
            return true
        }
        return false
    }

    /**
     * Call by the activity that holds the fragment if the back button is override.
     * If the function returns true, the back button is override to go up.
     * Else, it behaves normally.
     *
     * @return True if we can navigate up from the current directory. False otherwise.
     */
    fun canNavigateUp(): Boolean {

        if (currentFileInfo == null) {
            return false
        }

        val absolutePath = currentFileInfo?.absoluteFilePath

        return absolutePath != null &&
                absolutePath != root &&
                absolutePath.contains(File.separator)

    }

    /**
     * Calls navigateTo on the parent directory.
     */
    protected fun doNavigateUp() {
        val parentPath = FileUtils.truncatePath(currentFileInfo!!.absoluteFilePath)
        navigateTo(parentPath)
    }

    interface Callback {
        fun onDirectoryLoaded(dirContent: RemoteCommand.FileInfo)
    }

    companion object {
        @JvmField
        val PREVIOUS_DIRECTORY_PATH = ".."
        @JvmField
        val KEY_DIRECTORY_CONTENT = "DIRECTORY_CONTENT"
        @JvmField
        val DEFAULT_DIRECTORY_PATH = "default_path"
    }
}
